src/Gaffer/ContextVariablesTweaks.cpp

```diff
@@ -0,0 +1,122 @@
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "Gaffer/ContextVariablesTweaks.h"
+
+#include "IECore/DataAlgo.h"
+
+using namespace IECore;
+using namespace Gaffer;
+
+GAFFER_NODE_DEFINE_TYPE( ContextVariablesTweaks );
+
+size_t ContextVariablesTweaks::g_firstPlugIndex = 0;
+
+ContextVariablesTweaks::ContextVariablesTweaks( const std::string &name ) : ContextProcessor( name )
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+	addChild( new BoolPlug( "ignoreMissing", Plug::In, false ) );
+	addChild( new TweaksPlug ( "tweaks" ) );
+}
+
+ContextVariablesTweaks::~ContextVariablesTweaks()
+{
+
+}
+
+Gaffer::BoolPlug *ContextVariablesTweaks::ignoreMissingPlug()
+{
+	return getChild<BoolPlug>( g_firstPlugIndex );
+}
+
+const Gaffer::BoolPlug *ContextVariablesTweaks::ignoreMissingPlug() const
+{
+	return getChild<BoolPlug>( g_firstPlugIndex );
+}
+
+Gaffer::TweaksPlug *ContextVariablesTweaks::tweaksPlug()
+{
+	return getChild<TweaksPlug>( g_firstPlugIndex + 1 );
+}
+
+const Gaffer::TweaksPlug *ContextVariablesTweaks::tweaksPlug() const
+{
+	return getChild<TweaksPlug>( g_firstPlugIndex + 1 );
+}
+
+bool ContextVariablesTweaks::affectsContext( const Plug *input ) const
+{
+	return tweaksPlug()->isAncestorOf( input ) || input == ignoreMissingPlug();
+}
+
+void ContextVariablesTweaks::processContext( Context::EditableScope &context, IECore::ConstRefCountedPtr &storage ) const
+{
+	const TweaksPlug *tweaksPlug = this->tweaksPlug();
+	if( tweaksPlug->children().empty() )
+	{
+		return;
+	}
+
+	const bool ignoreMissing = ignoreMissingPlug()->getValue();
+
+	IECore::CompoundDataPtr sourceData = new CompoundData();
+	IECore::CompoundDataPtr result = new CompoundData();
+
+	tweaksPlug->applyTweaks(
+		[&context, &sourceData]( const std::string &valueName )
+		{
+			DataPtr value = context.context()->getAsData( valueName, nullptr );
+			sourceData->writable()[valueName] = value;
```

Comment:
`applyTweaks` doesn't take ownership of what we return here, which I presume is why `sourceData` is used to keep `value` alive. But `applyTweaks` also doesn't specify how long it needs us to keep the data alive. In practice it's only until it has finished applying the current tweak, which means until the next call to this lambda. So we could avoid the additional overhead (allocations, map lookups) of the `CompoundData`, and just store the single latest returned value in `ConstDataPtr sourceData`.

Alternatively, if we want to be pedantic and say that it's our responsibility to keep all the source data alive until after `applyTweaks` has completed, then I think technically we have a bug here. That's because a second tweak with the same `valueName` will overwrite `writable()[valueName]` with _different_ data, and the previously returned value will be destroyed. This could be fixed simply by storing `vector<ConstDataPtr> sourceData` and always pushing the value onto it rather than using a map. But I'd be inclined to go with the cheaper `ConstDataPtr sourceData` that we know should work in practice.

Reply:
Thanks for that explanation, you're right about the intention being to maintain the lifetimes. I went with your first recommendation in https://github.com/GafferHQ/gaffer/pull/5047/commits/19418b4fa6215cc4afabe83fbd08f1c8acb24563 as it makes sense to me that it does the job and keeps storage to a minimum.